### PR TITLE
contribute-to-tidb: update how to review a pr

### LINKED ISSUE #27: Integrate Review Best Practices among TiDB Community
So far, we have a first version on How to review a PR.

* https://zz-jason.gitbook.io/tidb-dev-guide/contribute-to-tidb/review-a-pr

However, it is most inspired by the content of Flink and Rust Community. We should continuously integrate Best Practices among TiDB Community.

Recently @tiancaiamao gives an offline talk about this topic. I'd like to invite him to join this effort.

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What issue does this PR solve?

- close #27 

### What is changed:

Add more content about PR review, and some of it are more TiDB specific.

### CODE REVIEW

**Comment by tisonkun** [comment]
Thanks for your pull request @tiancaiamao ! I'll review it today.

File: contribute-to-tidb/review-a-pr.md
Comment (djshow832) [inline contribute-to-tidb/review-a-pr.md:29]: It's changed to `Approve`, not `/LGTM`. Will you mention the method to approve the PR?

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:15]:
```suggestion
* Eliminate bugs early
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:18]:
```suggestion
Some high level guidelines about code review in the TiDB community:
```

Yep. It could be guidelines in PingCAP. But they are also contributors in TiDB community and actually how TiDB community works. We don't have to emphasize "PingCAP" here.

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:11]: Given that we don't use a checklist any more. It's better to remove the redundant `no.` prefix
```suggestion
## Things to know before the code review
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:69]:
```suggestion
## What to look for in a code review
```

ditto

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:141]:
```suggestion
## How to write code review comments
```

ditto

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:166]:
```suggestion
## Review with the @ti-chi-bot
```

ditto

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:22]: Nice statement!

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [reply contribute-to-tidb/review-a-pr.md:29]: Yes. Please use Approval

```suggestion
### 2 Approvals for a PR to be accepted
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:31]:
```suggestion
For most of the repositories in the TiDB community, we require 2 approvals before the PR is accepted. A few repositories require different number of approvals, but 2 approvals is the default setting. 2 approvals ensures that even after the original PR author's resignation, there is someone familiar with the code and the project still keeps moving. And if the author really resigns, the code reviewer is the first person responsible for the code.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:33]:
```suggestion
### Never give approval to to the code you don't understand
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:44]:
```suggestion

For example, a bugfix should be in a separated PR, a refactoring should be in a separate PR. No matter how small it is, like updating the package version, or even one line changes to fix a typo, the change does exactly one thing and gets it right.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:46]:
```suggestion
### Principles of the code review
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:49]:
```suggestion
* Software design is about trade-off and there is no silver bullet
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:53]:
```suggestion
A common mistake is to focus too much on style. Style is important, that is why there are style guides. Just let the bot do most of the trivial work. The code reviewer is a human, who is supposed to focus on things beyond a bot can do. They must make sure that the logic is not broken by the changes, and the maintenance cost of the code base is not decreased.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:55]:
```suggestion
Sometimes, making choices can be tricky.

Said that there is a special optimization that can improve the overall performance by 30%. However, it introduces a totally different code path and every feature later on has to consider this code path specially, which means a huge maintenance burden. Shall we accept or reject it?

Said that there is a critical bug and a PR is read to fix it, but the change is risky to introduce other kind of bugs. Shall we accept or reject it?
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:57]:
```suggestion
The answer is always “it depends”. Software design is more like a kind of art than technology. It is about aesthetics and your taste of the code. Trade-offs are always there and there is no perfect solution sometimes. Still, there is a general standard (from [@shenli](https://github.com/shenli), maybe we can call it "better than before?"):
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:66]:
```suggestion

For a bug fix PR, if it comes with no test case, the reviewer should ask the author to add tests.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:67]:
```suggestion

For a performance improvement PR, if no benchmark result is provided, the reviewer should ask the author to attach a benchmark result.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:27]: We can rephrase this section in "I" form instead of "he" for warming reading experience and avoid risk of sexism.

```suggestion
We had a fierce debate on how to speed up the review process long long ago. What the role the reviewer and PR owner should play and what is each of their responsibility to get the PR merged faster.

After the meeting, most of the participants agree with that the PR owner is responsible for getting PR merged.

If I am the PR author, I am the one who wants to merge the PR into the repository most. I'm willing to lead that change. So I'm supposed to do my best to have someone review the PR. The reviewers should review the PR in time and give feedback, but it is not their responsibility to push the PR to be merged.
```

**Review by tisonkun — CHANGES_REQUESTED** [review]
Thanks for your contribution @tiancaiamao ! It is really an excellent work! I left comments for "Things to know before the code review" section. Will give reviews on other section later.

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:105]:
```suggestion
Does the developer pick good names for everything? A good name is long enough to fully communicate what the item is or does, without being so long that it becomes hard to read.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:111]:
```suggestion
* Does the developer write clear comments in understandable English?
* Are all of the comments actually necessary?

Usually comments are useful when they explain why the code exists, while not explain what the code is doing. If the code isn't clear enough to explain itself, then the code should be made simpler.

There are some exceptions; for example, regular expressions and complex algorithms often benefit greatly from comments that explain what they’re doing. But mostly comments are for information that the code itself can hardly contain, like the reason behind a decision.

It can also be helpful to look at comments existing before this PR. Maybe there is a `TODO` can be removed now, a comment advising against this change being made, etc.

Note that comments are different from documentation of classes, modules, or functions, which should instead express the purpose of a piece of code, how it should be used, and how it behaves when used.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:117]:
```suggestion
Make sure the PR follows our [code style and quality guide](code-style-and-quality-guide.md). For Go and Rust, there are built in tools with the compiler toolchain.

If you want to improve style points outside the style guide, you can comment that you think it would improve the code readability but isn't mandatory. Don't block PRs from being submitted based only on personal style preferences.

The PR author should not include major style changes combined with other changes. For example, if a PR wants to reformat the whole file should not make functional changes. This has been explained as "One PR just does one thing" in the previous section.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:135]:
```suggestion
If a PR changes how users build, test, interact with, or release code, the reviewer are supposed to check whether it also updates related documentation, such as READMEs, and any generated reference docs. 

If a PR deletes or deprecates code, it should consider whether the documentation should also be deleted.

If documentation is missing, the reviewer should ask for it.

#### Error Handling

Be careful with error or exception handling. This is one of the most common places for bugs.

* Does the test cover the error code path?
* Does a panic make the whole process exit for server services?
* Does the code handle the resource releasing properly?
* Especially, for code including concurrent goroutines, is there a goroutine leak or memory leak after the error?

Sometimes it is difficult for a real program to run into the error code path, and even more difficult to reproduce and debug if it really happens. Then it is necessary to mock the error code path. Tools like [failpoint](https://github.com/pingcap/failpoint) are available for that purpose.
```

**Review by tisonkun — COMMENTED** [review]
... and the second section.

I do fix a few render issue and expression. But I think it should be reviewed after merged by content expert like @qiancai . We don't block the PR but the content contains too many Chinglish as well as imperative sentences.

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:145]:
```suggestion
Good reviewers are compassionate. Your code always need to be reviewed. And also you always need to review other contributors' code. Regard reviews as a learning process, and everyone wins.
```

File: contribute-to-tidb/review-a-pr.md
Comment (tisonkun) [inline contribute-to-tidb/review-a-pr.md:153]: Definitely!

File: contribute-to-tidb/review-a-pr.md
Comment (tiancaiamao) [reply contribute-to-tidb/review-a-pr.md:18]: Okay... writing much more PRs than reviewing is common in our company, especially for the new employees.
For the community contributors, they don't have such strong feeling.

Comment (tisonkun) [reply contribute-to-tidb/review-a-pr.md:18]: At least for this sentence, it emphasizes the importance of review itself. I don't feel an impact on "writing much more PRs than reviewing" among these words.

**Review by tisonkun — APPROVED** [review]
Thanks for your contribution @tiancaiamao ! Excellent content. Will do a grammar improvement later.

Merging...

**Comment by tisonkun** [comment]
@all-contributors please add @tiancaiamao  for content

**Comment by tisonkun** [comment]
@all-contributors please add @djshow832  for review

**Pull request merged**
